Initial can-bind implementation

chasenlehara opened the pull request:
This is the initial implementation required for can-stache-bindings, can-route, and can-component.

[Docs on the 5.0 site](https://canjs.github.io/next/doc/can-bind.html)

Docs on GitHub:

- [can-bind](https://github.com/canjs/can-bind/blob/initial-implementation/doc/can-bind.md)
- [parentValue](https://github.com/canjs/can-bind/blob/initial-implementation/doc/parentValue.md)
- [start()](https://github.com/canjs/can-bind/blob/initial-implementation/doc/start.md)
- [startChild()](https://github.com/canjs/can-bind/blob/initial-implementation/doc/startChild.md)
- [startParent()](https://github.com/canjs/can-bind/blob/initial-implementation/doc/startParent.md)
- [stop()](https://github.com/canjs/can-bind/blob/initial-implementation/doc/stop.md)
- [updateChild()](https://github.com/canjs/can-bind/blob/initial-implementation/doc/updateChild.md)
- [updateParent()](https://github.com/canjs/can-bind/blob/initial-implementation/doc/updateParent.md)

Part of https://github.com/canjs/canjs/issues/3738

justinbmeyer commented:
We typically have a small example in here: 

![image](https://user-images.githubusercontent.com/78602/41561142-ae75ce42-730e-11e8-9c69-b3b9eff9c1cb.png)

justinbmeyer commented:
![image](https://user-images.githubusercontent.com/78602/41561151-b6258b3c-730e-11e8-930b-29872c0827da.png)


I'm not sure this description will explain to a new user why they would use this.  What does "bind" mean?  What type of observables?

Maybe something like:

> Updates one observable value with the value of another observable.

justinbmeyer commented:
![image](https://user-images.githubusercontent.com/78602/41561260-08785400-730f-11e8-96df-3b4972f9a835.png)


To match parameters, I tend to use __bold__ for option names, and a `code` for type:

- __child__ `{ObservableValue}` _required_: The child observable.
- __ childToParent__ `{Boolean}` _optional_:  By default, can-bind will check if the child has the can.getValue symbol and either setParent is provided or the parent has the can.setValue symbol; providing this option overrides those checks with the option’s value (e.g. false will force the binding to be one-way parent-to-child).

justinbmeyer commented:
Why did we make `queue` required?

Do we want to make shorthand?

```js
new Bind( value.bind(ob1, "name"), value.bind(ob2, "fullName") )
```

justinbmeyer commented:
One thing we should mention, is that `can-bind` sets up all the appropriate debugging instrumentation.

justinbmeyer commented:
The `# Use` section should probably give an example w/i a component.  As much of a real-world use case as possible.

Bitballs's `time` might work:

```
time: "number",
newStat: {Default: Stat},
connectedCallback(){
  var bind = new Bind( 
    value.from(this, "time"),
    value.to(this,"newStat.time")
  );
  bind.start();
  return () => {
    bind.stop()
  }
}
```

chasenlehara commented:
Updated my PR:

- Made the queue option… optional (defaults to domUI)
- Moved the Overview section into [the signature](https://canjs.github.io/next/doc/can-bind.html#newBind_options_)
- Added sections on [debugging](https://canjs.github.io/next/doc/can-bind.html#Debugging) and [warnings](https://canjs.github.io/next/doc/can-bind.html#Warnings)
- Added a [“Use” section](https://canjs.github.io/next/doc/can-bind.html#Use) with an example of using can-bind with can-component
- Updated the formatting of the options in the signature
- Updated the package’s description

chasenlehara commented:
Merging this as-is; I’ll create another PR with more changes.
